File: src/agent/tools.py
import os
import subprocess
from typing import Dict, Any, Callable, List
from pydantic import BaseModel, Field, ValidationError

# ------------------------------------------------------------------------------
# Core Tool Interfaces
# ------------------------------------------------------------------------------
class ToolCall(BaseModel):
    """Structured tool call from LLM"""
    id: str = Field(default_factory=lambda: f"call_{id(object())}", description="Unique ID of the tool call")
    name: str = Field(description="Name of the tool to call")
    parameters: Dict[str, Any] = Field(description="Parameters for the tool")

class ToolResult(BaseModel):
    """Standardized tool execution result"""
    success: bool = Field(description="Whether the tool executed successfully")
    content: str = Field(description="Output content or error message")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="Additional metadata")

class BaseTool:
    """Abstract base class for all tools"""
    name: str
    description: str
    parameters: Dict[str, Any]

    def __init__(self, project_root: str):
        self.project_root = os.path.abspath(project_root)
        # Ensure project root exists
        if not os.path.exists(self.project_root):
            raise ValueError(f"Project root {self.project_root} does not exist")

    def run(self, parameters: Dict[str, Any]) -> ToolResult:
        """Execute the tool with given parameters"""
        raise NotImplementedError("Subclasses must implement run()")

    def to_schema(self) -> Dict[str, Any]:
        """Return tool schema for LLM function calling"""
        # Extract required list and remove 'required' from property definitions
        required = [k for k, v in self.parameters.items() if v.get("required")]
        properties = {k: {kk: vv for kk, vv in v.items() if kk != "required"}
                      for k, v in self.parameters.items()}

        return {
            "type": "function",
            "function": {
                "name": self.name,
                "description": self.description,
                "parameters": {
                    "type": "object",
                    "properties": properties,
                    "required": required
                }
            }
        }

# ------------------------------------------------------------------------------
# Core Tool Implementations
# ------------------------------------------------------------------------------
class ReadFileTool(BaseTool):
    name = "read_file"
    description = "Read the contents of a file from the project directory"
    parameters = {
        "file_path": {
            "type": "string",
            "description": "Relative path to the file from project root",
            "required": True
        }
    }

    def _validate_path(self, file_path: str) -> str:
        """Validate path is within project root and resolve absolute path"""
        abs_path = os.path.abspath(os.path.join(self.project_root, file_path))
        root = os.path.join(self.project_root, "")  # ensure trailing separator
        if not abs_path.startswith(root):
            raise PermissionError(f"Path traversal detected: {file_path} is outside project root")
        return abs_path

    def run(self, parameters: Dict[str, Any]) -> ToolResult:
        try:
            file_path = parameters["file_path"]
            abs_path = self._validate_path(file_path)

            if not os.path.exists(abs_path):
                return ToolResult(success=False, content=f"File not found: {file_path}")
            if not os.path.isfile(abs_path):
                return ToolResult(success=False, content=f"Path is not a file: {file_path}")

            file_size = os.path.getsize(abs_path)
            if file_size > 10 * 1024 * 1024:  # 10 MB limit
                return ToolResult(success=False, content=f"File too large: {file_path} ({file_size} bytes, max 10MB)")

            with open(abs_path, "r", encoding="utf-8") as f:
                content = f.read()

            return ToolResult(
                success=True,
                content=content,
                metadata={"file_path": file_path, "size": len(content)}
            )
        except Exception as e:
            return ToolResult(success=False, content=f"Error reading file: {str(e)}")

class WriteFileTool(BaseTool):
    name = "write_file"
    description = "Write content to a file in the project directory (overwrites existing files)"
    parameters = {
        "file_path": {
            "type": "string",
            "description": "Relative path to the file from project root",
            "required": True
        },
        "content": {
            "type": "string",
            "description": "Content to write to the file",
            "required": True
        }
    }

    def _validate_path(self, file_path: str) -> str:
        """Validate path is within project root and resolve absolute path"""
        abs_path = os.path.abspath(os.path.join(self.project_root, file_path))
        root = os.path.join(self.project_root, "")  # ensure trailing separator
        if not abs_path.startswith(root):
            raise PermissionError(f"Path traversal detected: {file_path} is outside project root")
        return abs_path

    def run(self, parameters: Dict[str, Any]) -> ToolResult:
        try:
            file_path = parameters["file_path"]
            content = parameters["content"]
            abs_path = self._validate_path(file_path)

            # Create parent directories if they don't exist
            os.makedirs(os.path.dirname(abs_path), exist_ok=True)

            with open(abs_path, "w", encoding="utf-8") as f:
                f.write(content)

            return ToolResult(
                success=True,
                content=f"Successfully wrote {len(content)} characters to {file_path}",
                metadata={"file_path": file_path, "size": len(content)}
            )
        except Exception as e:
            return ToolResult(success=False, content=f"Error writing file: {str(e)}")

class RunShellTool(BaseTool):
    name = "run_shell"
    description = (
        "Run a shell command in the project directory. "
        "Allowed commands: ls, cat, grep, find, git, python, python3, npm, node, "
        "pwd, echo, head, tail, wc, sort, uniq, cut, tr, sed, awk, xargs, diff, "
        "mkdir, cp, mv, touch, dirname, basename, which, tee, printf, env. "
        "Pipes (|) and output redirection (>, >>) are allowed. "
        "Operators ;, &, $, and backticks are blocked."
    )
    parameters = {
        "command": {
            "type": "string",
            "description": "Shell command to execute (see tool description for allowed commands and operators)",
            "required": True
        }
    }

    ALLOWED_COMMANDS = {
        "ls", "cat", "grep", "find", "git", "python", "python3", "npm", "node",
        "pwd", "echo", "head", "tail", "wc", "sort", "uniq", "cut", "tr", "sed",
        "awk", "xargs", "diff", "mkdir", "cp", "mv", "touch", "dirname", "basename",
        "which", "tee", "printf", "env",
    }

    # Dangerous commands — blocked as the base command (first word).
    BLOCKED_COMMANDS = {"rm", "sudo", "su", "chmod", "chown", "wget", "curl", "ssh", "scp", "eval"}

    # Shell metacharacters blocked only when they appear unquoted.
    # Mapped to: (multi_char_variant, still_dangerous_in_double_quotes)
    SHELL_META = {
        ";":  (None, False),
        "&":  (None, False),
        "$":  (None, True),
        "`":  (None, True),
    }

    def _scan_unquoted(self, command: str) -> str | None:
        """Walk the command tracking shell quote state.

        Returns the first dangerous metacharacter found outside an
        appropriate quoting context, or None if the command is clean.
        """
        in_single = False
        in_double = False
        i = 0
        while i < len(command):
            ch = command[i]

            if ch == "'" and not in_double:
                in_single = not in_single
                i += 1
                continue
            if ch == '"' and not in_single:
                in_double = not in_double
                i += 1
                continue

            if ch == '\\' and not in_single:
                i += 2  # skip escaped char
                continue

            # Double-quoted: only $ and ` remain dangerous
            if in_double and ch in ('$', '`'):
                return ch

            if not in_single and not in_double:
                meta = self.SHELL_META.get(ch)
                if meta:
                    multi, _ = meta
                    if multi and command.startswith(multi, i):
                        return multi
                    return ch

            i += 1
        return None

    def _validate_command(self, command: str) -> None:
        """Validate command is safe to execute"""
        cmd_parts = command.strip().split()
        if not cmd_parts:
            raise ValueError("Empty command")

        base_cmd = cmd_parts[0]
        if base_cmd not in self.ALLOWED_COMMANDS:
            raise PermissionError(
                f"Command '{base_cmd}' is not allowed. "
                f"Allowed commands: {', '.join(sorted(self.ALLOWED_COMMANDS))}"
            )
        if base_cmd in self.BLOCKED_COMMANDS:
            raise PermissionError(f"Command '{base_cmd}' is blocked")

        offender = self._scan_unquoted(command)
        if offender:
            raise PermissionError(f"Command contains blocked shell operator: {offender}")

    def run(self, parameters: Dict[str, Any]) -> ToolResult:
        try:
            command = parameters["command"]
            self._validate_command(command)

            result = subprocess.run(
                command,
                shell=True,
                cwd=self.project_root,
                capture_output=True,
                text=True,
                timeout=30  # 30 second timeout
            )

            if result.returncode == 0:
                output = result.stdout
                success = True
            else:
                output = f"Command failed with exit code {result.returncode}\nStderr: {result.stderr}"
                success = False

            return ToolResult(
                success=success,
                content=output,
                metadata={"command": command, "returncode": result.returncode}
            )
        except Exception as e:
            return ToolResult(success=False, content=f"Error running command: {str(e)}")

# ------------------------------------------------------------------------------
# Tool Registry
# ------------------------------------------------------------------------------
class ToolRegistry:
    """Registry for managing and executing tools"""
    def __init__(self, project_root: str):
        self.project_root = project_root
        self.tools: Dict[str, BaseTool] = {}
        # Register core tools
        self.register_tool(ReadFileTool(project_root))
        self.register_tool(WriteFileTool(project_root))
        self.register_tool(RunShellTool(project_root))

    def register_tool(self, tool: BaseTool) -> None:
        """Register a new tool"""
        self.tools[tool.name] = tool

    def get_tool_schemas(self) -> List[Dict[str, Any]]:
        """Get all tool schemas for LLM function calling"""
        return [tool.to_schema() for tool in self.tools.values()]

    def execute_tool_call(self, tool_call: ToolCall) -> ToolResult:
        """Execute a tool call"""
        if tool_call.name not in self.tools:
            return ToolResult(
                success=False,
                content=f"Tool {tool_call.name} not found. Available tools: {', '.join(self.tools.keys())}"
            )

        tool = self.tools[tool_call.name]
        return tool.run(tool_call.parameters)

File: tui/src/agent/tools.ts
import * as fs from "node:fs";
import * as path from "node:path";
import { execSync } from "node:child_process";

// ── types ────────────────────────────────────────────────────────────────

export interface ToolCall {
  id: string;
  name: string;
  parameters: Record<string, unknown>;
}

export interface ToolResult {
  success: boolean;
  content: string;
  metadata: Record<string, unknown>;
}

export interface ToolSchema {
  type: "function";
  function: {
    name: string;
    description: string;
    parameters: {
      type: "object";
      properties: Record<string, { type: string; description: string }>;
      required: string[];
    };
  };
}

// ── abstract base ────────────────────────────────────────────────────────

export abstract class BaseTool {
  abstract name: string;
  abstract description: string;
  abstract parameters: Record<string, { type: string; description: string; required?: boolean }>;

  constructor(protected projectRoot: string) {
    this.projectRoot = path.resolve(projectRoot);
    if (!fs.existsSync(this.projectRoot)) {
      throw new Error(`Project root ${this.projectRoot} does not exist`);
    }
  }

  abstract run(parameters: Record<string, unknown>): ToolResult | Promise<ToolResult>;

  toSchema(): ToolSchema {
    const required: string[] = [];
    const properties: Record<string, { type: string; description: string }> = {};

    for (const [key, val] of Object.entries(this.parameters)) {
      if (val.required) required.push(key);
      const { required: _, ...rest } = val;
      properties[key] = rest;
    }

    return {
      type: "function",
      function: {
        name: this.name,
        description: this.description,
        parameters: { type: "object", properties, required },
      },
    };
  }

  protected validatePath(filePath: string): string {
    const abs = path.resolve(this.projectRoot, filePath);
    const root = this.projectRoot.endsWith(path.sep)
      ? this.projectRoot
      : this.projectRoot + path.sep;
    if (!abs.startsWith(root)) {
      throw new Error(`Path traversal detected: ${filePath} is outside project root`);
    }
    return abs;
  }
}

// ── ReadFile ─────────────────────────────────────────────────────────────

export class ReadFileTool extends BaseTool {
  name = "read_file";
  description = "Read the contents of a file from the project directory";
  parameters = {
    file_path: {
      type: "string",
      description: "Relative path to the file from project root",
      required: true,
    },
  };

  run(params: Record<string, unknown>): ToolResult {
    try {
      const filePath = params.file_path as string;
      const abs = this.validatePath(filePath);

      if (!fs.existsSync(abs)) return { success: false, content: `File not found: ${filePath}`, metadata: {} };
      const stat = fs.statSync(abs);
      if (!stat.isFile()) return { success: false, content: `Path is not a file: ${filePath}`, metadata: {} };
      if (stat.size > 10 * 1024 * 1024) {
        return { success: false, content: `File too large: ${filePath} (${stat.size} bytes, max 10MB)`, metadata: {} };
      }

      const content = fs.readFileSync(abs, "utf-8");
      return { success: true, content, metadata: { file_path: filePath, size: content.length } };
    } catch (e: any) {
      return { success: false, content: `Error reading file: ${e.message}`, metadata: {} };
    }
  }
}

// ── WriteFile ────────────────────────────────────────────────────────────

export class WriteFileTool extends BaseTool {
  name = "write_file";
  description = "Write content to a file in the project directory (overwrites existing files)";
  parameters = {
    file_path: {
      type: "string",
      description: "Relative path to the file from project root",
      required: true,
    },
    content: {
      type: "string",
      description: "Content to write to the file",
      required: true,
    },
  };

  run(params: Record<string, unknown>): ToolResult {
    try {
      const filePath = params.file_path as string;
      const content = params.content as string;
      const abs = this.validatePath(filePath);

      fs.mkdirSync(path.dirname(abs), { recursive: true });
      fs.writeFileSync(abs, content, "utf-8");

      return {
        success: true,
        content: `Successfully wrote ${content.length} characters to ${filePath}`,
        metadata: { file_path: filePath, size: content.length },
      };
    } catch (e: any) {
      return { success: false, content: `Error writing file: ${e.message}`, metadata: {} };
    }
  }
}

// ── RunShell ─────────────────────────────────────────────────────────────

export class RunShellTool extends BaseTool {
  name = "run_shell";
  description =
    "Run a shell command in the project directory. " +
    "The user will be prompted to approve each command before it executes. " +
    "Pipes (|) and output redirection (>, >>) are allowed.";
  parameters = {
    command: {
      type: "string",
      description: "Shell command to execute (user must approve before execution)",
      required: true,
    },
  };

  run(params: Record<string, unknown>): ToolResult {
    try {
      const command = (params.command as string).trim();
      if (!command) return { success: false, content: "Empty command", metadata: {} };

      const result = execSync(command, {
        cwd: this.projectRoot,
        timeout: 30000,
        encoding: "utf-8",
        stdio: ["pipe", "pipe", "pipe"],
      });

      return { success: true, content: result, metadata: { command, returncode: 0 } };
    } catch (e: any) {
      const stderr = e.stderr || "";
      const status = e.status ?? 1;
      return {
        success: false,
        content: `Command failed with exit code ${status}\nStderr: ${stderr}`,
        metadata: { command: params.command as string, returncode: status },
      };
    }
  }
}

// ── ToolRegistry ─────────────────────────────────────────────────────────

export class ToolRegistry {
  private tools = new Map<string, BaseTool>();

  constructor(projectRoot: string) {
    this.register(new ReadFileTool(projectRoot));
    this.register(new WriteFileTool(projectRoot));
    this.register(new RunShellTool(projectRoot));
  }

  register(tool: BaseTool): void {
    this.tools.set(tool.name, tool);
  }

  getToolSchemas(): ToolSchema[] {
    return [...this.tools.values()].map((t) => t.toSchema());
  }

  async executeToolCall(toolCall: ToolCall): Promise<ToolResult> {
    const tool = this.tools.get(toolCall.name);
    if (!tool) {
      return {
        success: false,
        content: `Tool ${toolCall.name} not found. Available: ${[...this.tools.keys()].join(", ")}`,
        metadata: {},
      };
    }
    return await tool.run(toolCall.parameters);
  }
}

// ── SearchCodebase ─────────────────────────────────────────────────────────

import type { HybridSearch } from "../rag/hybrid-search.js";

export class SearchCodebaseTool extends BaseTool {
  name = "search_codebase";
  description =
    "Search the project codebase using hybrid (embedding + BM25) retrieval. " +
    "Use this when you need to find relevant code — functions, classes, interfaces, " +
    "or files — related to a specific concept, symbol, or question. " +
    "Results include the chunk type, file path, line range, and code content.";
  parameters = {
    query: {
      type: "string",
      description: "The search query — describe what you're looking for in natural language",
      required: true,
    },
    max_results: {
      type: "number",
      description: "Maximum results to return (default: 5, max: 20)",
      required: false,
    },
  };

  private hybrid: HybridSearch;

  constructor(projectRoot: string, hybrid: HybridSearch) {
    super(projectRoot); // projectRoot not used by this tool, but BaseTool requires it
    this.hybrid = hybrid;
  }

  async run(params: Record<string, unknown>): Promise<ToolResult> {
    try {
      const query = (params.query as string)?.trim();
      if (!query) {
        return { success: false, content: "Query cannot be empty", metadata: {} };
      }
      const maxResults = Math.min(Math.max(1, (params.max_results as number) ?? 5), 20);

      const results = await this.hybrid.search(query);

      if (results.length === 0) {
        return { success: true, content: `No results found for: "${query}"`, metadata: { query, count: 0 } };
      }

      const lines: string[] = [
        `Search results for: "${query}"`,
        `Found ${results.length} relevant code chunks:\n`,
      ];

      for (const r of results.slice(0, maxResults)) {
        const f = r.chunk.filePath.split("/").pop() ?? r.chunk.filePath;
        const parent = r.chunk.parentName ? ` (in ${r.chunk.parentName})` : "";
        const sig = r.chunk.signature ? `\n       Signature: ${r.chunk.signature.split("\n")[0]}` : "";
        lines.push(
          `  ── ${r.chunk.type} ${r.chunk.name}${parent} ──`,
          `     File: ${f}:${r.chunk.startLine}-${r.chunk.endLine}`,
          `     Relevance: ${(r.score * 100).toFixed(0)}%${sig}`,
          `     Code:`,
          ...r.chunk.content.split("\n").slice(0, 15).map((l) => `       │ ${l}`),
        );
        if (r.chunk.content.split("\n").length > 15) {
          lines.push(`       │ … (${r.chunk.content.split("\n").length - 15} more lines)`);
        }
        lines.push("");
      }

      return {
        success: true,
        content: lines.join("\n"),
        metadata: { query, count: results.length },
      };
    } catch (e: any) {
      return { success: false, content: `Error searching codebase: ${e.message ?? String(e)}`, metadata: {} };
    }
  }
}
